File: bigData/hw7/MATLAB/insurance.m
%% ================== Reset the workspace ===================
clear all;close all;clc;

%% ================== Insurance Dataset =====================
[data,txt,raw] = xlsread('insurance.xls','data');
y = double(strcmp(txt(2:end,31),'Yes')); % 0 and 1
y(y==0)=-1; % 1 and -1
X = [data(2:end, 1:7),data(2:end, 9:end)]; % only numeric fields
X = normalize(X); % binary fields are not affected from min-max normalization
[m,n] = size(X);
beta = 0.95;

% Stratified sampling, 2 strata
train_pc = 0.7;
train_ind = [randsample(1:sum(y==-1),floor(train_pc*sum(y==-1))),sum(y==-1)+ ...
             randsample(1:sum(y==1),floor(train_pc*sum(y==1)))];
TrainSize = size(train_ind,2);
yTrain = y(ismember(1:m,train_ind),:);
yTest = y(~ismember(1:m,train_ind),:);
XTrain = X(ismember(1:m,train_ind),:);
XTest = X(~ismember(1:m,train_ind),:);

% New representation of the training and test data sets
[XTrain,w] = pca(XTrain,beta);
XTest=XTest*w;

% Gaussian Kernel
sigma = 0.1;
K = kernel(XTrain,XTrain,sigma,0);
a0=eps*ones(TrainSize,1);

% Inequality that individual alpha>=0
A = -eye(TrainSize);
b = zeros(TrainSize,1);

% Equality that sum(alpha_i*y_i)=0
Aeq = yTrain';
beq = 0;

C = 1e+20;

% Change from min to max optimization by multiplying with -1
% Regularization term to force H positive definite
H = diag(yTrain)*K*diag(yTrain)+1e-10*eye(TrainSize);
f = -ones(TrainSize,1);

options = optimset('UseParallel','always','Display','iter','MaxIter',100, ...
                   'MaxFunEvals',1000000,'LargeScale', 'off');
alpha = quadprog(H,f,A,b,Aeq,beq,zeros(TrainSize,1),C*ones(TrainSize,1),a0,options);

indx = find(alpha >= eps);
b = mean(yTrain(indx)-K(indx,:)*(alpha.*yTrain));
pred = sign(((1 + kernel(XTest, XTrain, sigma, 0)) .^ 2) * ...
            (alpha .* yTrain) + b);
fprintf('Test Accuracy: %f\n', mean(double(pred==yTest))*100);

pred = sign(((1+kernel(XTrain,XTrain,sigma,0)).^2)*(alpha.*yTrain)+b);
fprintf('Train Accuracy: %f\n', mean(double(pred==yTrain))*100);

File: bigData/hw7/R/insurance.R
#!/usr/bin/Rscript


## attach package for solving quadratic programming
library(quadprog)

## attach functions we write
source("functions.R")

## set up data path for possibly different OS
datPath <- paste("..", "data", "insurance.csv", sep = .Platform$file.sep)
dat <- read.csv(datPath)            # "dat" used since `data` is a function in R
y <- ifelse(dat[, 31L] == "Yes", 2L, 1L)
X <- as.matrix(dat[, sapply(dat[1:10, ], class) %in% c("integer", "numeric")])
X <- normalize(X)                       # min-max normalization
m <- nrow(X)
n <- ncol(X)
beta <- 0.95

## prepare training and test set
set.seed(123)                    # random seed is important for reproducibility

## stratified sampling with 2 strata
train_pc <- 0.7
train_ind <- c(sample(which(y == 1L), floor(train_pc * sum(y == 1L))),
               sample(which(y == 2L), floor(train_pc * sum(y == 2L))))
TrainSize <- length(train_ind)
yTrain <- y[train_ind]
yTest <- y[- train_ind]
XTrain <- X[train_ind, ]
XTest <- X[- train_ind, ]


## new representation of the training and test dataset
tmpList <- pca(XTrain, beta)
XTrain <- tmpList$pca
XTest <- XTest %*% tmpList$w


## Gaussian Kernel
sigma <- 0.1
K <- kernel(XTrain, XTrain, sigma, 0)

## Inequality that individual alpha >= 0
A <- diag(TrainSize)                # different with MATLAB code due to solve.QP
b <- rep(0, TrainSize)

## Equality that sum(alpha_i * y_i) = 0
Aeq <- yTrain
beq <- 0

## Change from min to max optimization by multiplying with - 1
## Regularization term to force H positive definite
H <- diag(as.vector(yTrain)) %*% K %*% diag(as.vector(yTrain)) +
    1e-10 * diag(TrainSize)
f <- - rep(1, TrainSize)

## call function solve.QP
res <- solve.QP(Dmat = H, dvec = f, Amat = cbind(Aeq, A),
                bvec = c(beq, b), meq = 1L)

## correct prediction rate of training set
indx <- which(res$solution >= .Machine$double.eps)
bb <- colMeans(yTrain[indx] - K[indx, ] %*% (res$solution * yTrain))

## correct prediction rate on test set
pred = sign(((1 + kernel(XTest, XTrain, sigma, 0)) ^ 2) %*%
            (res$solution * yTrain) + bb);
message(sprintf("\nTest Set Accuracy: %f\n",
                mean(as.integer(pred) == yTest) * 100))

## correct prediction rate on training set
pred = sign(((1 + kernel(XTrain, XTrain, sigma, 0)) ^ 2) %*%
            (res$solution * yTrain) + bb);
message(sprintf("\nTrain Set Accuracy: %f\n",
                mean(as.integer(pred) == yTrain) * 100))
